plugins/urls/__init__.py

```diff
@@ -1,165 +1,228 @@
+
+    def reload(self):
         self.shortened.runQuery("CREATE TABLE IF NOT EXISTS urls ("
                                 "url TEXT, "
                                 "shortener TEXT, "
                                 "result TEXT)")
 
-        self.catcher = Catcher(self, self.config, self.storage, self.logger)
+        for handler_list in self.handlers.itervalues():
+            for handler in handler_list:
+                handler.reload()
 
-        self.blacklist = []
-        blacklist = self.config.get("blacklist", [])
+    @inlineCallbacks
+    def shorten(self, _url, shortener=None, target=None):
+        if isinstance(_url, basestring):
+            match = extract_urls(_url)
 
-        for element in blacklist:
-            try:
-                self.blacklist.append(re.compile(element))
-            except Exception:
-                self.logger.exception("Unable to compile regex '%s'" % element)
+            if len(match) < 1:
+                return
 
-    def check_blacklist(self, url):
-        """
-        Check whether a URL is in the user-defined blacklist
+            match = match[0]
 
-        :param url: The URL to check
-        :type url: str
+            self.logger.trace("match: {0}", match)
+            # Expand the match to make it easier to work with
 
-        :return: Whether the URL is in the blacklist
-        :rtype: bool
-        """
+            # Input: ''http://x:y@tools.ietf.org:80/html/rfc1149''
+            # Match: '', http, x:y, tools.ietf.org, :80, /html/rfc1149''
+            _prefix, _protocol, _basic, _domain, _port, _path = (
+                to_unicode(x) for x in match
+            )
+
+            _port = _port.lstrip(":")
 
-        for pattern in self.blacklist:
             try:
-                self.logger.debug(_("Checking pattern '%s' against URL '%s'")
-                                  % (pattern, url))
-                if re.match(pattern, url):
-                    return True
-            except Exception as e:
-                self.logger.debug(_("Error in pattern matching: %s") % e)
-                return False
-        return False
+                if _port:
+                    _port = int(_port)
+            except ValueError:
+                self.logger.warn("Invalid port: {0}", _port)
+                return
+
+            _url = URL(self, _protocol, _basic, _domain, _port, _path)
+
+        if target is not None:
+            shortener = self.get_shortener(target)
+
+        if shortener is None:
+            shortener = self.default_shortener
+
+        context = {"url": _url}
+
+        r = yield self.shortened.runQuery(
+            "SELECT * FROM urls WHERE url=? AND shortener=?",
+            (_url.text, shortener.lower())
+        )
 
-    @run_async_threadpool
-    def message_handler(self, event=MessageReceived):
+        if isinstance(r, Failure):
+            returnValue(r)
+        elif len(r):
+            returnValue(r[0][2])
```

Comment:
Is the `[2]` to get the 3rd column? If so, perhaps a `SELECT result` would be ebtter than a `SELECT *` above.